### DISCUSSION TITLE
RFC: A better name for `TaskResult`

### DISCUSSION THREAD

**Discussion by RealOrangeOne** [RealOrangeOne/django-tasks#19] [Q&A]
The name of `TaskResult` was copied from Celery, which uses a similar name for the handle to a given task. This naming however could be confusing, as it in itself isn't the "result" of the task (that's at `.result`).

Therefore, is there a better and more descriptive name for this?

Yes, naming things is hard!

**Comment by mgax** [comment]
The model is currently `DBTaskResult` which I think has a couple of issues:
- `DB` is implied because it's part of the `django_tasks_database` app.
- `Result` is somewhat misleading, because the model doesn't only represent a task's result, but also its arguments, routing information and status.

To be honest I'd vote for the short and simple name `Task`.

**Reply by mgax** [reply]
I've just noticed there's already a `django_tasks.task.Task` class so naming the model `Task` would certainly cause confusion. Perhaps `DBTask` would not be too bad then :)

**Reply by RealOrangeOne** [reply]
`DBTaskResult` is the `DB` version of a `TaskResult` - it's the `TaskResult` name I don't like. Sticking a `DB` prefix on whatever we call it is a fine name IMO.

I agree calling it a "result" when it's not really a result is a little confusing. But oddly, Celery seems to do this just fine. RQ deals with this by not having an equivilent for `Task`, and calling the running item a `Job`. I know DLQ uses `Task` and `Job` for the config and running item. I like that model, but the 2 are used by some people interchangeably.

**Comment by hooverdc** [comment]
`TaskInstance`, as in you've enqueued an instance of that task?

**Reply by RealOrangeOne** [reply]
The issue with "instance" is that it already has a definition, and gets a bit confusing and odd to talk about. "A `TaskInstance`" and "A `Task` instance" sound identical. Not to mention "A `TaskInstance` instance" sounds a little heavy-handled

**Reply by hooverdc** [reply]
Ah, good points.

**Comment by jhthompson** [comment]
What about `TaskRun`? Since it seems like this class represents a single "run" of a particular "task"?

**Reply by hooverdc** [reply]
`TaskExecution` if you want to be wordier.

**Reply by jhthompson** [reply]
Could work too! I like the terseness of `TaskRun`, but that might just be me.

**Reply by hooverdc** [reply]
On a reread I like `TaskRun` more than my suggestion.

**Reply by jhthompson** [reply]
Made a PR for this change here: https://github.com/RealOrangeOne/django-tasks/pull/101 to gauge interest!

**Comment by RealOrangeOne** [comment]
Perhaps the correct solution here is actually to flip it. Instead, rename `TaskResult` to `Task`, and find a better name for `Task`. It feels more natural to me to ask "Did that task run?" where "task" is the thing which ran, as opposed to the configuration for it.

Perhaps the `task` decorator returns a new type subclassing `Task`, as opposed to an instance of a `Task`, so "Task instance" becomes a sensible name again, and somewhat matches the behaviour of models.

**Reply by mgax** [reply]
That sounds sensible! Perhaps `Task` can be `TaskDefinition` or `TaskConfiguration`.

**Reply by jhthompson** [reply]
I like this!

What about the `@task` decorator? Are we okay with the disconnect between the naming of that decorator now returning a `TaskConfiguration` instead of a `Task`?

We could do:
```python
@configure_task
def do_the_thing()
    return None
```

or

```python
@task_configuration
def do_the_thing()
    return None
````

But I think I would be fine with leaving it as `@task`, it's a lot cleaner.

**Reply by RealOrangeOne** [reply]
That would be a breaking change, but I suspect you're right that the naming should be consistent.

I'll have a think and a play later today on sensible namings and refactoring.

**Reply by RealOrangeOne** [reply]
I don't think "TaskDefinition" works, as you end up with "A definition of a TaskDefinition", which is a little heavy-handed.

My intention had been to be inspired by some other queue libraries. However, most either don't distinguish between the 2, use ambiguous words (eg "Task" and "Job"), or use similar names to what there is now :sweat_smile:

**Reply by RealOrangeOne** [reply]
I suspect not distinguishing is the way to go. The "Task" is the definition, and it's an instance of it which is run. Changing the `@task` to be a class itself rather than an instance of `Task` might help cement that, and allows for much more customization in future. 
I'm going to have a play around.

**Reply by jhthompson** [reply]
Yeah maybe not distinguishing is the way to go... Just out of curiosity after you mentioned it, I went through a bunch of other library/frameworks and tried to figure out what they called these things:

| Library | "definition" / "configuration" | "instance" |
|---------|--------------|------------|
|     [celery](https://github.com/celery/celery/tree/main)    |      [Task](https://github.com/celery/celery/blob/b1061a0f8eddd7df2c321d3ccad532b35f31992a/celery/app/task.py#L164)     |      [Result](https://github.com/celery/celery/blob/b1061a0f8eddd7df2c321d3ccad532b35f31992a/celery/worker/request.py#L67)      |
|    [huey](https://github.com/coleifer/huey/tree/master)     |       [Task](https://github.com/coleifer/huey/blob/895f63556ccd3679650eba6186dcb9bc19d25b85/huey/api.py#L680)       |      [Result](https://github.com/coleifer/huey/blob/895f63556ccd3679650eba6186dcb9bc19d25b85/huey/api.py#L937)      |
|    [laravel](https://github.com/laravel/laravel)     |       [Job](https://laravel.com/docs/11.x/queues#class-structure)       |      N/A      |
| [rq](https://github.com/rq/rq) | any function / [@job](https://github.com/rq/rq/blob/d5e55ab95a9d22978963ec5b19509380559b654a/rq/decorators.py#L15) | [Job](https://github.com/rq/rq/blob/d5e55ab95a9d22978963ec5b19509380559b654a/rq/job.py#L144) |
| [sidekiq](https://github.com/sidekiq/sidekiq) | [Job](https://github.com/sidekiq/sidekiq/blob/ed8b217d32f026acc1fae7221dca18d1baa86d11/lib/sidekiq/job.rb#L44) | N/A |

**Comment by RealOrangeOne** [comment]
I've addressed one side of this in https://github.com/RealOrangeOne/django-tasks/pull/103, but I'm still not completely happy with the naming.
